auth0/src/main/java/com/auth0/android/provider/BrowserPicker.java

```diff
@@ -0,0 +1,160 @@
+        dest.writeStringList(allowedPackages);
+    }
+
+    @Override
+    public int describeContents() {
+        return 0;
+    }
+
+    public static final Creator<BrowserPicker> CREATOR = new Creator<BrowserPicker>() {
+        @Override
+        public BrowserPicker createFromParcel(Parcel in) {
+            return new BrowserPicker(in);
+        }
+
+        @Override
+        public BrowserPicker[] newArray(int size) {
+            return new BrowserPicker[size];
+        }
+    };
+
+    /**
+     * Create a new BrowserPicker.Builder instance.
+     *
+     * @return a new BrowserPicker.Builder ready to customize.
+     */
+    @NonNull
+    public static BrowserPicker.Builder newBuilder() {
+        return new BrowserPicker.Builder();
+    }
+
+    public static class Builder {
+        private List<String> allowedPackages;
+
+        private Builder() {
+        }
+
+        //TODO: JAVADOCS
+        @NonNull
+        public Builder withAllowedPackages(@NonNull List<String> allowedPackages) {
+            this.allowedPackages = allowedPackages;
+            return this;
+        }
+
+        @NonNull
+        public BrowserPicker build() {
+            return new BrowserPicker(allowedPackages);
+        }
+
+    }
+
+    @Nullable
+    String getBestBrowserPackage(@NonNull PackageManager pm) {
+        Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse("http://www.example.com"));
+        ResolveInfo webHandler = pm.resolveActivity(browserIntent,
+                Build.VERSION.SDK_INT >= Build.VERSION_CODES.M ? PackageManager.MATCH_ALL : PackageManager.MATCH_DEFAULT_ONLY);
+        String defaultBrowser = null;
+        if (webHandler != null) {
+            defaultBrowser = webHandler.activityInfo.packageName;
+        }
+
+        final List<ResolveInfo> availableBrowsers = pm.queryIntentActivities(browserIntent, 0);
+        final List<String> regularBrowsers = new ArrayList<>();
+        final List<String> customTabsBrowsers = new ArrayList<>();
+        final boolean isFilterEnabled = allowedPackages != null;
+
+        for (ResolveInfo info : availableBrowsers) {
+            boolean isAllowed = !isFilterEnabled || allowedPackages.contains(info.activityInfo.packageName);
+            if (!isAllowed) {
+                continue;
+            }
+            Intent serviceIntent = new Intent();
+            serviceIntent.setAction(ACTION_CUSTOM_TABS_CONNECTION);
+            serviceIntent.setPackage(info.activityInfo.packageName);
+            if (pm.resolveService(serviceIntent, 0) != null) {
+                customTabsBrowsers.add(info.activityInfo.packageName);
+            } else {
+                regularBrowsers.add(info.activityInfo.packageName);
+            }
+        }
+
+        //If the browser packages were filtered, use the allowed packages list as order preference.
+        //A user-selected default browser will always be picked up first.
+        List<String> preferenceList = allowedPackages;
```

Comment:
the order of preference is dictated by either the "allowed packages" in the case that the developer has provided those, or the default browser + chrome browsers, in the case of the legacy scenario